mkosi/kmod.py

```diff
@@ -380,25 +380,34 @@ def gen_required_kernel_modules(
         # /usr/lib/firmware/nvidia/tu106/gsp
         # /usr/lib/firmware/nvidia/tu117/gsp
 
-        for p in reversed(path_and_parents_below(context.root / fw, context.root / firmwared)):
-            # Make sure only the last component of the path we're working is a symlink. We don't want to
-            # yield any paths with multiple components in them that are symlinks as these paths are passed to
-            # cpio and it interprets intermediary path components that are symlinks as regular directories.
-            # Additionally, intermediary path components that are symlinks also mess up the path
-            # deduplication we want to get by putting all the firmware to add in a set as the paths will hash
-            # differently even though they're the same paths after resolution.
-            p = p.parent.resolve(strict=True).joinpath(p.name)
+        todo = list(reversed(fw.parts))
+        current = context.root
+        while todo:
+            part = todo.pop()
+            if part == ".":
+                continue
+            elif part == "..":
+                current = current.parent
+                continue
+
+            current /= part
+            if not current.is_symlink():
+                continue
+
+            firmware.add(current.relative_to(context.root))
 
-            while p.is_symlink():
-                target = p.readlink()
-                if target.is_absolute():
-                    die(f"Found unexpected absolute symlink {target} in {firmwared}")
+            target = current.readlink()
+            parts = list(target.parts)
```

Comment:
Making the tuple a list here shouldn't be necessary
```suggestion
            parts = target.parts
```